# RNA-seq_AVC
this pipeline takes fastq files oalready aligned data from NGI-RNAseq pipeline and calls allele specific variants

[![Build Status](https://travis-ci.org/rna-seq_AVC.svg?branch=master)](https://travis-ci.org/rna-seq_AVC)
[![Nextflow](https://img.shields.io/badge/nextflow-%E2%89%A50.24.0-brightgreen.svg)](https://www.nextflow.io/)


### Introduction
RNA-seq_AVC: this pipeline takes fastq files oalready aligned data from NGI-RNAseq pipeline and calls allele specific variants

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It comes with docker / singularity containers making installation trivial and results highly reproducible.


### Documentation
The RNA-seq_AVC pipeline comes with documentation about the pipeline, found in the `docs/` directory:

1. [Installation](docs/installation.md)
2. Pipeline configuration
    * [Local installation](docs/configuration/local.md)
    * [Adding your own system](docs/configuration/adding_your_own.md)
3. [Running the pipeline](docs/usage.md)
4. [Output and how to interpret the results](docs/output.md)
5. [Troubleshooting](docs/troubleshooting.md)

### Credits
This pipeline was written by Aron T. Skaftason ([arontommi](https://github.com/arontommi)) at [Your Institute](http://www.yourinstitute.com).
